### PR TITLE
[RTM] allow to tag service as data container callback

### PR DESCRIPTION
**Pull request by aschempp**
Same as with hooks, you can now tag services as DCA callbacks.

Example tags: 
 - `{ name: contao.callback, table: tl_page, target: config.onload }`
 - `{ name: contao.callback, table: tl_member_group, target: list.label.label }`
 - `{ name: contao.callback, table: tl_article, target: fields.inColumn.options }`
 - `{ name: contao.callback, table: tl_content, target: fields.headline.save, method: mySaveMethod }`

Same as with hooks, the method is automatically generated if not given. E.g. `onLoadCallback` or `onOptionsCallback`.

**ToDo:**
 - [x] Unit Tests

### CODE REVIEW

**Comment by Toflar** [comment]
This looks really nice!

File: core-bundle/src/EventListener/DataContainerCallbackListener.php [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                $value = [$key => $value];
+            }
+
+            $replaces[] = $value;
+        }
+
+        $GLOBALS['TL_DCA'][$table] = array_replace_recursive($GLOBALS['TL_DCA'][$table], $replaces);
+    }
+
+    private function getFromDCA(string $table, array $keys)
```
Comment (leofeyer) [inline core-bundle/src/EventListener/DataContainerCallbackListener.php]: `: array` can be added IMHO

**Comment by leofeyer** [comment]
I like this one. Implementation looks good to me, so as soon as you have added the unit tests, this one is ready to merge IMHO.

File: core-bundle/src/EventListener/DataContainerCallbackListener.php (hunk 1 above)
Comment (aschempp) [reply core-bundle/src/EventListener/DataContainerCallbackListener.php]: No, it can be `null`, `array` or `callable` (a closure)

Comment (leofeyer) [reply core-bundle/src/EventListener/DataContainerCallbackListener.php]: Then you should add a phpDoc `@return`. 😉

Comment (aschempp) [reply core-bundle/src/EventListener/DataContainerCallbackListener.php]: Added in 15bd74688c5c3c3ceb26f201f982159ffd99a458

**Comment by aschempp** [comment]
Unit tests for the listener are still missing, but everything else added. I was unable to test the alias handling, I'm not sure how this works. An alias cannot have tags, so why is this method needed at all (I've copied it from the `RegisterHookListenersPass`)?

**Comment by aschempp** [comment]
I've now added all remaining unit tests

**Review by ausi — CHANGES_REQUESTED** [review]
I proposed some changes to the `DataContainerCallbackListener` in aschempp/contao#1

**Comment by aschempp** [comment]
I have merged the suggestions of @ausi, I think this is RTM now.

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Pull request merged**